### PR TITLE
accessing hardware resources through controller interface

### PR DESCRIPTION
**Pull request by Karsten1987**
Signed-off-by: Karsten Knese <Karsten1987@users.noreply.github.com>

### CODE REVIEW

File: design_drafts/hardware_access.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+The hardware is composed uniquely through the URDF.
+
+**Joint**
+A joint is considered a logical component and is being actuated by at least one actuator (it might be generically under- or over-actuated depending on the actual hardware setup).
+The joint is meant to be abstraction layer between a high-level controller instance and the underlaying hardware.
+The abstraction is needed to shim over a potentially complex hardware setup in order to control a joint.
+A single joint might be controlled by multiple motors with a non-trivial transmission interface, yet a controller only really thinks in joint state values, not in transmission ratios.
+
+A joint is configured in conjunction to a hardware resource such as **Actuator**, **Sensor** or **System**.
+The joint component is configured through command and state interfaces.
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:20]:
```suggestion
The joint component is used through command and state interfaces which are declared in the URDF.
```

File: design_drafts/hardware_access.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+The command interfaces describe the value in which this joint can be controlled (e.g. effort or velocity) where as the state interfaces are considered the state feedback interfaces.
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:21]:
```suggestion
The command interfaces describe the value in which this joint can be controlled (e.g. effort or velocity) where as the state interfaces are considered read-only feedback interfaces.
```

File: design_drafts/hardware_access.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+<joint name="joint1">
+  <command_interface>velocity</command_interface>
+  <state_interface>position</state_interface>
+  <state_interface>velocity</state_interface>
+</joint>
+...
+```
+
+**Actuator**
+An actuator describes a single motor instance with at max 1DoF and is strictly tight to a **Joint**.
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:34]:
```suggestion
A standalone actuator component describes a single motor instance with at max 1DoF and is strictly tied to a **Joint**.
```

File: design_drafts/hardware_access.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+It might hereby take a single command value for its appropriate mode of operation, such as a desired joint velocity or effort - in rare cases an actuator might actually take a precise joint position value.
+The implementation of this actuator might then convert the desired value into PWM or other hardware specific commands and control the hardware.
+Similarly, an actuator might provide state feedback.
+Depending the setup, the motor encoders might provide position, velocity, effort or current feedback.
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:38]:
```suggestion
Depending on the setup, the motor encoders might provide position, velocity, effort or current feedback.
```

File: design_drafts/hardware_access.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+<ros2_control name="my_simple_servo_motor" type="actuator">
+<hardware>
+  <class>simple_servo_motor_pkg/SimpleServoMotor</class>
+  <param name="serial_port">/dev/tty0</param>
+...
+</hardware>
+  <joint name="base_link">
+  <command_interface>position</command_interface>
+  <state_interface>position</state_interface>
+  <param name="max_value">1.57</param>
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:52]: We clarified the min-max notation in [the urdf examples](https://github.com/ros-controls/roadmap/blob/master/design_drafts/components_architecture_and_urdf_examples.md) where max and min are child nodes of the command interface, the state interface shouldn't have limits on it. The notation you use in this example looks simpler and I'd personally prefer it if it wasn't prone to misinterpretation. Better be explicit in this case.

In short,
```
      <commandInterfaceType name="position">
        <param name="min">-1</param>
        <param name="max">1</param>
      </commandInterfaceType>
```

File: design_drafts/hardware_access.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+  <param name="min_value">-1.57<param>
+...
+</joint>
+</ros2_control>
+```
+The snippet above depicts a relative simple hardware setup, which a single actuator which controls one logical joint.
+The joint hereby is configured to be commanded in position values, whereas the state feedback is also position.
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:59]:
```suggestion
The joint here is configured to be commanded in position values, whereas the state feedback is also position.
```

File: design_drafts/hardware_access.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  </sensor>
+</ros2_control>
+```
+Note that we technically have a separation between a physical hardware resource (`<hardware type="sensor">`) and a logical component (`<sensor name="my_sensor">`), both called *Sensor*.
+We don't individually specify them further in this document as they don't have a significant semantic interpretation for the user.
+
+**System**
+A system is meant to be a more complex hardware setup which contains multiple joints and sensors.
+This is mostly used for third-party robotic systems such as robotic arms or industrial setups, which have their own (proprietary) API.
+The implementation of the system hardware resource thus serves as an interface between the hardware and the controller to provide this API.
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:88]:
```suggestion
The implementation of the system hardware resource serves as an interface between the hardware and the controller to provide this API.
```

File: design_drafts/hardware_access.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+      <command_interface>effort</command_interface>
+      <state_interface>position</state_interface>
+      <state_interface>velocity</state_interface>
+      <state_interface>effort</state_interface>
+    </joint>
+</ros2_control>
+```
+
+## Resource Manager
+The resource manager is responsible for parsing the URDF and instantiating the respective hardware resources.
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:113]:
```suggestion
The resource manager is responsible for parsing the URDF and instantiating the respective hardware components.
```

File: design_drafts/hardware_access.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+It has ownership over the lifetime of these hardware resources and their associated logical joints.
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:114]:
```suggestion
It has ownership over the lifetime of these hardware components and their associated logical joints.
```

File: design_drafts/hardware_access.md [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+It serves as the storage backend for the controller manager which can loan resources to the controller.
+The resource manager hereby administrates which controllers claimed which resources.
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:116]:
```suggestion
The resource manager keeps a ledger of controllers and their respective claimed resources.
```

File: design_drafts/hardware_access.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+If a controller does no longer need access to the claimed resource, that resource is being released and offered for other controllers to be claimed.
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:117]:
```suggestion
If a controller no longer needs access to the claimed resource, it is released and returns to the ResourceManager where it may be offered for other controllers.
```

File: design_drafts/hardware_access.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+The resource manager internally maintains a mapping of each individual hardware resources and their interfaces.
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:119]:
```suggestion
The resource manager internally maintains a mapping of each individual hardware resource and their interfaces.
```

File: design_drafts/hardware_access.md [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+This mapping can be indexed through a simple `<logical_component>/<interface_name>` lookup.
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:120]: I spent a few seconds thinking this may refer to some XML, may be fortunate to use a different style

File: design_drafts/hardware_access.md [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+As mentioned previously it abstracts the individual hardware resources from their logical components, such that a user (or controller in that sense) does not have to know which hardware is responsible for commanding a joint.
+
+In the examples above, the actuator command interfaces are being mapped to `base_link/position`, the state interfaces equivalently to `base_link/position`.
+Likewise for the sensor, their interfaces are being mapped to `my_sensor/roll`, `my_sensor/pitch`, `my_sensor/yaw`.
+
+## Controller Interface
+Once the system is bootstrapped and a controller is loaded, it can claim logical components and access their interfaces.
+
+**Generic Access**
+A controller has the chance to access a single interface value by directly indexing the resource manager for the respective key, such as `joint1/effort`, which allows to set effort values during the controllers execution.
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:130]:
```suggestion
A controller has the chance to access a single interface value via a query to the resource manager for the respective key, such as `joint1/effort`, which - if available - claims the handle that allows to set `effort` values on `joint1` during the execution of this controller.
```

File: design_drafts/hardware_access.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+```c++
+void MyController::init(ResourceManager * resource_manager)
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:132]: A code snippet here is unfortunate as now we can start debating implementation approaches. One could argue that this is only a reference, the problem with that is they are sticky and then may be used later as justification. Please forgive me :D 

No raw pointers please. 
A simple change:
```suggestion
void MyController::init(... resource_manager)
```

File: design_drafts/hardware_access.md [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+  InterfaceCommandHandle joint1_effort_cmd = resource_manager->claim_command_interface("joint1/effort");
+  InterfaceStateHandle joint1_position_state = resource_manager->claim_state_interface("joint1/position");
+}
+```
+
+**Semantic Components**
+While the above example might suffice for simple setups, one can imagine that there might be quite some handles accumulated, e.g. when dealing with 6D FT/Sensors or IMUs.
+We therefore propose semantic components which wrap a non-zero amount of keys and provide a more meaningful API on top of it.
+```c++
+void MyController::init(ResourceManager * resource_manager)
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:143]:
```suggestion
void MyController::init(... resource_manager)
```

File: design_drafts/hardware_access.md [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+{
+  FTSensor6D ft_sensor(resource_manager,
+    "sensor1/fx", "sensor1/fy", "sensor1/fz",  // force values
+    "sensor1/tx", "sensor1/ty", "sensor1/tz"); // torque values
+
+  std::vector<double> torque = ft_sensor.get_torque_values();
```
Comment (bmagyar) [inline design_drafts/hardware_access.md:149]:
```suggestion
  std::vector<double> torque = ft_sensor.get_torque_values();
  geometry_msgs::msg::Wrench wrench_msg = ft_sensor.as_wrench_msg();
```

**Review by bmagyar — CHANGES_REQUESTED** [review]

**Review by destogl — CHANGES_REQUESTED** [review]
Very nice work!

I mostly propose simplification of some sentences to provide straightforward guidelines.

File: design_drafts/hardware_access.md [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+}
+```
+
+**Semantic Components**
+While the above example might suffice for simple setups, one can imagine that there might be quite some handles accumulated, e.g. when dealing with 6D FT/Sensors or IMUs.
+We therefore propose semantic components which wrap a non-zero amount of keys and provide a more meaningful API on top of it.
+```c++
+void MyController::init(ResourceManager * resource_manager)
+{
+  FTSensor6D ft_sensor(resource_manager,
```
[the remark this answers is not in the document]
Comment (Karsten1987) [reply design_drafts/hardware_access.md:145]: Maybe that wasn't clear, but the example here with the `FTSensor6D` is that convenience component.

File: design_drafts/hardware_access.md [hunk 19]
```diff
@@ -0,0 +1,160 @@
+# Hardware Access through Controllers
+
+The following article describe how controllers can exclusively claim hardware resources yet stay as flexible as possible to avoid a strongly typed triple of 'position', 'velocity' and 'effort [c.f. [Flexible Joint States Message](https://github.com/ros-controls/roadmap/blob/master/design_drafts/flexible_joint_states_msg.md)].
+We firstly describe how hardware resources are classified and loaded.
+Secondly, we provide a real-time safe way of accessing these resources through controllers.
+
+## Hardware Resources
+A hardware resource describes a physical component which is being considered through ros2-control.
+We hereby distinguish between three classes of hardware resources, namely Actuators, Sensor and System.
+Every individual hardware is loaded at runtime and thus allows a flexible and dynamic composition of the to be controlled setup.
```
[the remark this answers is not in the document]
Comment (bmagyar) [reply design_drafts/hardware_access.md:10]:
```suggestion
Each individual hardware is loaded at runtime and thus allows a flexible and dynamic composition of the to-be-controlled setup.
```

**Review by bmagyar — APPROVED** [review]
There are some outstanding points but I think this is mergeable nevertheless.

File: design_drafts/hardware_access.md [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+  <state_interface>position</state_interface>
+  <state_interface>velocity</state_interface>
+</joint>
+...
+```
+
+**Actuator**
+An actuator describes a single motor instance with at max 1DoF and is strictly tight to a **Joint**.
+It might hereby take a single command value for its appropriate mode of operation, such as a desired joint velocity or effort - in rare cases an actuator might actually take a precise joint position value.
+The implementation of this actuator might then convert the desired value into PWM or other hardware specific commands and control the hardware.
```
[the remark this answers is not in the document]
Comment (Karsten1987) [reply design_drafts/hardware_access.md:36]: I prefer the original text here. Otherwise I feel like it reads as it has to be converted to PWM. That's just an example.

**Review by destogl — APPROVED** [review]
It looks good!

File: design_drafts/hardware_access.md [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+<ros2_control name="my_simple_servo_motor" type="actuator">
+  <hardware>
+    <class>simple_servo_motor_pkg/SimpleServoMotor</class>
+    <param name="serial_port">/dev/tty0</param>
+  ...
+  </hardware>
+  <joint name="joint1">
+    <command_interface name="position">
+      <param name="max_value">1.57</param>
+      <param name="min_value">-1.57<param>
```
Comment (destogl) [inline design_drafts/hardware_access.md:70]:
```suggestion
      <param name="min">-1.57<param>
      <param name="max">1.57</param>
```

File: design_drafts/hardware_access.md [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+```xml
+<ros2_control name="my_simple_servo_motor" type="actuator">
+  <hardware>
+    <class>simple_servo_motor_pkg/SimpleServoMotor</class>
+    <param name="serial_port">/dev/tty0</param>
+  ...
+  </hardware>
+  <joint name="joint1">
+    <command_interface name="position">
+      <param name="min">-1.57<param>
```
Comment (Karsten1987) [inline design_drafts/hardware_access.md:69]:
```suggestion
      <param name="min">-1.57</param>
```

**Comment by Karsten1987** [comment]
With both reviews approved, going ahead and merge.
